File: apps/web/lib/stores/pipeline.ts
import { create } from 'zustand'
import { PipelineRun, PipelineStep, SSEEvent } from '@/types'

interface PipelineStore {
  runs: PipelineRun[]
  currentRun: PipelineRun | null
  loading: boolean
  error: string | null
  
  // Actions
  setRuns: (runs: PipelineRun[]) => void
  setCurrentRun: (run: PipelineRun | null) => void
  updateRunStep: (runId: string, step: PipelineStep) => void
  updateRunStatus: (runId: string, status: any) => void
  handleSSEEvent: (event: SSEEvent) => void
  setLoading: (loading: boolean) => void
  setError: (error: string | null) => void
}

export const usePipelineStore = create<PipelineStore>((set, get) => ({
  runs: [],
  currentRun: null,
  loading: false,
  error: null,

  setRuns: (runs) => set({ runs }),
  
  setCurrentRun: (run) => set({ currentRun: run }),
  
  updateRunStep: (runId, step) =>
    set((state) => ({
      runs: state.runs.map((run) =>
        run.id === runId
          ? {
              ...run,
              steps: run.steps.map((s) => (s.id === step.id ? step : s)),
            }
          : run
      ),
      currentRun:
        state.currentRun?.id === runId
          ? {
              ...state.currentRun,
              steps: state.currentRun.steps.map((s) =>
                s.id === step.id ? step : s
              ),
            }
          : state.currentRun,
    })),

  updateRunStatus: (runId, status) =>
    set((state) => ({
      runs: state.runs.map((run) =>
        run.id === runId ? { ...run, ...status } : run
      ),
      currentRun:
        state.currentRun?.id === runId
          ? { ...state.currentRun, ...status }
          : state.currentRun,
    })),

  handleSSEEvent: (event) => {
    const { event: eventType, data, run_id, step_id } = event

    switch (eventType) {
      case 'step_update':
        if (step_id && data) {
          const step: PipelineStep = {
            id: step_id,
            run_id,
            step_name: data.step_name,
            status: data.status,
            started_at: data.started_at,
            completed_at: data.completed_at,
            agent_reasoning: data.agent_reasoning,
            input_artifact_key: data.input_artifact_key,
            output_artifact_key: data.output_artifact_key,
            metadata: data.metadata || {},
            error_message: data.error_message,
          }
          get().updateRunStep(run_id, step)
        }
        break

      case 'run_completed':
      case 'run_failed':
        get().updateRunStatus(run_id, {
          status: data.status,
          completed_at: data.completed_at,
          error_message: data.error_message,
        })
        break
    }
  },

  setLoading: (loading) => set({ loading }),
  
  setError: (error) => set({ error }),
}))

File: apps/api/schemas/pipeline.py
from pydantic import BaseModel, ConfigDict, Field
from typing import Optional, Dict, Any, List
from datetime import datetime
from models.pipeline_run import PipelineStatus, StepStatus, StepName
import uuid


class PipelineConfigStep(BaseModel):
    step_name: StepName
    enabled: bool = True
    params: Dict[str, Any] = Field(default_factory=dict)


class PipelineRunCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    target_sequence: str = Field(..., min_length=10)
    pipeline_config: List[PipelineConfigStep]


class PipelineStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    id: uuid.UUID
    run_id: uuid.UUID
    step_name: StepName
    status: StepStatus
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    agent_reasoning: Optional[str]
    input_artifact_key: Optional[str]
    output_artifact_key: Optional[str]
    metadata: Dict[str, Any]
    error_message: Optional[str]


class PipelineRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    id: uuid.UUID
    user_id: uuid.UUID
    name: str
    status: PipelineStatus
    target_sequence: str
    pipeline_config: List[Dict[str, Any]]
    created_at: datetime
    completed_at: Optional[datetime]
    error_message: Optional[str]
    run_key: Optional[str]
    steps: List[PipelineStepResponse] = Field(default_factory=list)


class PipelineRunList(BaseModel):
    runs: List[PipelineRunResponse]
    total: int
    page: int
    per_page: int


class SSEEvent(BaseModel):
    event: str
    data: Dict[str, Any]
    run_id: str
    step_id: Optional[str] = None